// @quicknode/sdk/esm/src/api/controllers/contracts.js
import { __decorate, __metadata } from 'tslib';
import { DEFAULT_CHAIN } from '../utils/constants.js';
import { contractDetailsValidator } from '../types/contracts/getContractDetails.js';
import { CodegenEthMainnetContractDetailsDocument, CodegenPolygonMainnetContractDetailsDocument, CodegenEthSepoliaContractDetailsDocument } from '../graphql/generatedTypes.js';
import { ValidateInput } from '../../lib/validation/ValidateInput.js';

class ContractsController {
    constructor(client, defaultChain = DEFAULT_CHAIN) {
        this.client = client;
        this.defaultChain = defaultChain;
    }
    async getDetails(variables) {
        const { chain, ...queryVariables } = variables;
        const userChain = chain || this.defaultChain;
        const query = {
            ethereum: CodegenEthMainnetContractDetailsDocument,
            polygon: CodegenPolygonMainnetContractDetailsDocument,
            ethereumSepolia: CodegenEthSepoliaContractDetailsDocument,
        };
        const { data: { [userChain]: { contract }, }, } = await this.client.query({
            variables: queryVariables,
            query: query[userChain],
        });
        if (contract)
            return { contract };
        return { contract: null };
    }
}
__decorate([
    ValidateInput(contractDetailsValidator),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], ContractsController.prototype, "getDetails", null);

export { ContractsController };
