// @urql/exchange-graphcache/dist/urql-exchange-graphcache.mjs
import { stringifyVariables as e, formatDocument as r, createRequest as t, makeOperation as a, stringifyDocument as i } from "@urql/core";

import { Kind as n, valueFromASTUntyped as o } from "@0no-co/graphql.web";

import { share as s, map as v, filter as l, merge as d, mergeMap as u, empty as p, fromArray as c, makeSubject as f, onPush as y } from "wonka";

var h = "\nhttps://bit.ly/2XbVrpR#";

var m = new Set;

var g = [];

var popDebugNode = () => g.pop();

var pushDebugNode = (e, r) => {
  var t = "";
  if (r.kind === n.INLINE_FRAGMENT) {
    t = e ? `Inline Fragment on "${e}"` : "Inline Fragment";
  } else if (r.kind === n.OPERATION_DEFINITION) {
    t = `${r.name ? `"${r.name.value}"` : "Unnamed"} ${r.operation}`;
  } else if (r.kind === n.FRAGMENT_DEFINITION) {
    t = `"${r.name.value}" Fragment`;
  }
  if (t) {
    g.push(t);
  }
};

var getDebugOutput = () => g.length ? "\n(Caused At: " + g.join(", ") + ")" : "";

function invariant(e, r, t) {
  if (!e) {
    var a = r || "Minfied Error #" + t + "\n";
    if ("production" !== process.env.NODE_ENV) {
      a += getDebugOutput();
    }
    var i = new Error(a + h + t);
    i.name = "Graphcache Error";
    throw i;
  }
}

function warn(e, r) {
  if (!m.has(e)) {
    console.warn(e + getDebugOutput() + h + r);
    m.add(e);
  }
}

var _ = {};

var getDirectives = e => e._directives || _;

var getName = e => e.name.value;

var getFragmentTypeName = e => e.typeCondition.name.value;

var getFieldAlias = e => e.alias ? e.alias.value : e.name.value;

var N = [];

var getSelectionSet = e => e.selectionSet ? e.selectionSet.selections : N;

var getTypeCondition = e => e.typeCondition ? e.typeCondition.name.value : null;

var getFieldArguments = (e, r) => {
  var t = null;
  if (e.arguments) {
    for (var a = 0, i = e.arguments.length; a < i; a++) {
      var n = e.arguments[a];
      var s = o(n.value, r);
      if (null != s) {
        if (!t) {
          t = {};
        }
        t[getName(n)] = s;
      }
    }
  }
  return t;
};

var filterVariables = (e, r) => {
  if (!r || !e.variableDefinitions) {
    return;
  }
  var t = {};
  for (var a = 0, i = e.variableDefinitions.length; a < i; a++) {
    var n = getName(e.variableDefinitions[a].variable);
    t[n] = r[n];
  }
  return t;
};

var normalizeVariables = (e, r) => {
  var t = {};
  if (!r) {
    return t;
  }
  if (e.variableDefinitions) {
    for (var a = 0, i = e.variableDefinitions.length; a < i; a++) {
      var n = e.variableDefinitions[a];
      var s = getName(n.variable);
      t[s] = void 0 === r[s] && n.defaultValue ? o(n.defaultValue, r) : r[s];
    }
  }
  for (var v in r) {
    if (!(v in t)) {
      t[v] = r[v];
    }
  }
  return t;
};

function getMainOperation(e) {
  for (var r = 0; r < e.definitions.length; r++) {
    if (e.definitions[r].kind === n.OPERATION_DEFINITION) {
      return e.definitions[r];
    }
  }
  invariant(!1, "production" !== process.env.NODE_ENV ? "Invalid GraphQL document: All GraphQL documents must contain an OperationDefinitionnode for a query, subscription, or mutation." : "", 1);
}

var getFragments = e => {
  var r = {};
  for (var t = 0; t < e.definitions.length; t++) {
    var a = e.definitions[t];
    if (a.kind === n.FRAGMENT_DEFINITION) {
      r[getName(a)] = a;
    }
  }
  return r;
};

var shouldInclude = (e, r) => {
  var t = getDirectives(e);
  if (t.include || t.skip) {
    for (var a in t) {
      var i = t[a];
      if (i && ("include" === a || "skip" === a) && i.arguments && i.arguments[0] && "if" === getName(i.arguments[0])) {
        var n = o(i.arguments[0].value, r);
        return "include" === a ? !!n : !n;
      }
    }
  }
  return !0;
};

var isDeferred = (e, r) => {
  var {defer: t} = getDirectives(e);
  if (t) {
    for (var a of t.arguments || []) {
      if ("if" === getName(a)) {
        return !!o(a.value, r);
      }
    }
    return !0;
  }
  return !1;
};

var k = "__";

var isFieldNullable = (e, r, t) => {
  var a = getField(e, r, t);
  return !!a && "NON_NULL" !== a.type.kind;
};

var isListNullable = (e, r, t) => {
  var a = getField(e, r, t);
  if (!a) {
    return !1;
  }
  var i = "NON_NULL" === a.type.kind ? a.type.ofType : a.type;
  return "LIST" === i.kind && "NON_NULL" !== i.ofType.kind;
};

var isFieldAvailableOnType = (e, r, t) => 0 === t.indexOf(k) || 0 === r.indexOf(k) || !!getField(e, r, t);

var isInterfaceOfType = (e, r, t) => {
  if (!t) {
    return !1;
  }
  var a = getTypeCondition(r);
  if (!a || t === a) {
    return !0;
  } else if (e.types.has(a) && "OBJECT" === e.types.get(a).kind) {
    return a === t;
  }
  !function expectAbstractType(e, r) {
    invariant(e.types.has(r) && ("INTERFACE" === e.types.get(r).kind || "UNION" === e.types.get(r).kind), "production" !== process.env.NODE_ENV ? "Invalid Abstract type: The type `" + r + "` is not an Interface or Union type in the defined schema, but a fragment in the GraphQL document is using it as a type condition." : "", 5);
  }(e, a);
  expectObjectType(e, t);
  return e.isSubType(a, t);
};

var getField = (e, r, t) => {
  if (0 === t.indexOf(k) || 0 === r.indexOf(k)) {
    return;
  }
  expectObjectType(e, r);
  var a = e.types.get(r).fields()[t];
  if ("production" !== process.env.NODE_ENV) {
    if (!a) {
      warn("Invalid field: The field `" + t + "` does not exist on `" + r + "`, but the GraphQL document expects it to exist.\nTraversal will continue, however this may lead to undefined behavior!", 4);
    }
  }
  return a;
};

function expectObjectType(e, r) {
  invariant(e.types.has(r) && "OBJECT" === e.types.get(r).kind, "production" !== process.env.NODE_ENV ? "Invalid Object type: The type `" + r + "` is not an object in the defined schema, but the GraphQL document is traversing it." : "", 3);
}

function warnAboutResolver(e) {
  "production" !== process.env.NODE_ENV && warn(`Invalid resolver: \`${e}\` is not in the defined schema, but the \`resolvers\` option is referencing it.`, 23);
}

var keyOfField = (r, t) => t ? `${r}(${e(t)})` : r;

var joinKeys = (e, r) => `${e}.${r}`;

var fieldInfoOfKey = e => {
  var r = e.indexOf("(");
  if (r > -1) {
    return {
      fieldKey: e,
      fieldName: e.slice(0, r),
      arguments: JSON.parse(e.slice(r + 1, -1))
    };
  } else {
    return {
      fieldKey: e,
      fieldName: e,
      arguments: null
    };
  }
};

var deserializeKeyInfo = e => {
  var r = e.indexOf(".");
  return {
    entityKey: e.slice(0, r).replace(/%2e/g, "."),
    fieldKey: e.slice(r + 1)
  };
};

var E = null;

var O = null;

var w = null;

var b = null;

var D = null;

var x = null;

var V = !1;

var I = !1;

function makeData(e, r) {
  var t;
  if (e) {
    if (E.has(e)) {
      return e;
    }
    t = O.get(e);
  }
  if (null == t) {
    t = r ? [] : {};
  }
  if (e) {
    O.set(e, t);
  }
  E.add(t);
  return t;
}

var ownsData = e => !!e && E.has(e);

var initDataState = (e, r, t, a, i) => {
  E = new WeakSet;
  O = new WeakMap;
  D = e;
  w = r;
  x = new Set;
  I = !!a;
  V = !!i;
  if ("production" !== process.env.NODE_ENV) {
    g.length = 0;
  }
  if (!t) {
    b = null;
  } else if ("read" === D) {
    b = t;
  } else if (a || r.hydrating || r.optimisticOrder.length > 1) {
    if (!a && !r.commutativeKeys.has(t)) {
      reserveLayer(r, t);
    } else if (a) {
      if (-1 !== r.optimisticOrder.indexOf(t) && !r.commutativeKeys.has(t)) {
        r.optimisticOrder.splice(r.optimisticOrder.indexOf(t), 1);
      }
      r.commutativeKeys.delete(t);
    }
    b = t;
    createLayer(r, t);
  } else {
    b = null;
    deleteLayer(r, t);
  }
};

var clearDataState = () => {
  if ("production" !== process.env.NODE_ENV) {
    getCurrentDependencies();
  }
  var e = w;
  var r = b;
  I = !1;
  b = null;
  if (!e.hydrating && r && e.optimisticOrder.indexOf(r) > -1) {
    var t = e.optimisticOrder.length;
    while (--t >= 0 && e.dirtyKeys.has(e.optimisticOrder[t]) && e.commutativeKeys.has(e.optimisticOrder[t])) {
      squashLayer(e.optimisticOrder[t]);
    }
  }
  E = null;
  O = null;
  D = null;
  w = null;
  x = null;
  if ("production" !== process.env.NODE_ENV) {
    g.length = 0;
  }
  if ("test" !== process.env.NODE_ENV) {
    if (!e.defer && (e.storage || !e.optimisticOrder.length)) {
      e.defer = !0;
      setTimeout((() => {
        initDataState("read", e, null);
        gc();
        persistData();
        clearDataState();
        e.defer = !1;
      }));
    }
  }
};

var noopDataState = (e, r, t) => {
  if (r && !t) {
    e.deferredKeys.delete(r);
  }
  initDataState("write", e, r, t);
  clearDataState();
};

var getCurrentDependencies = () => {
  invariant(null !== x, "production" !== process.env.NODE_ENV ? "Invalid Cache call: The cache may only be accessed or mutated duringoperations like write or query, or as part of its resolvers, updaters, or optimistic configs." : "", 2);
  return x;
};

var setNode = (e, r, t, a) => {
  if ("production" !== process.env.NODE_ENV) {
    invariant("read" !== D, "production" !== process.env.NODE_ENV ? "Invalid Cache write: You may not write to the cache during cache reads.  Accesses to `cache.writeFragment`, `cache.updateQuery`, and `cache.link` may  not be made inside `resolvers` for instance." : "", 27);
  }
  var i = b ? e.optimistic.get(b) : e.base;
  var n = i.get(r);
  if (void 0 === n) {
    i.set(r, n = Object.create(null));
  }
  if (void 0 === a && !b) {
    delete n[t];
  } else {
    n[t] = a;
  }
};

var getNode = (e, r, t) => {
  var a;
  var i = !I && "read" === D && b && w.commutativeKeys.has(b);
  for (var n = 0, o = w.optimisticOrder.length; n < o; n++) {
    var s = w.optimisticOrder[n];
    var v = e.optimistic.get(s);
    i = i && s !== b;
    if (v && (!i || !w.commutativeKeys.has(s)) && (!I || "write" === D || w.commutativeKeys.has(s)) && void 0 !== (a = v.get(r)) && t in a) {
      return a[t];
    }
  }
  return void 0 !== (a = e.base.get(r)) ? a[t] : void 0;
};

var updateRCForLink = (e, r) => {
  if (Array.isArray(e)) {
    for (var t = 0, a = e.length; t < a; t++) {
      updateRCForLink(e[t], r);
    }
  } else if ("string" == typeof e) {
    ((e, r) => {
      var t = w.refCount.get(e) || 0;
      var a = t + r > 0 ? t + r : 0;
      w.refCount.set(e, a);
      if (!a) {
        w.gc.add(e);
      } else if (!t && a) {
        w.gc.delete(e);
      }
    })(e, r);
  }
};

var extractNodeFields = (e, r, t) => {
  if (void 0 !== t) {
    for (var a in t) {
      if (!r.has(a)) {
        e.push(fieldInfoOfKey(a));
        r.add(a);
      }
    }
  }
};

var extractNodeMapFields = (e, r, t, a) => {
  extractNodeFields(e, r, a.base.get(t));
  for (var i = 0, n = w.optimisticOrder.length; i < n; i++) {
    var o = a.optimistic.get(w.optimisticOrder[i]);
    if (void 0 !== o) {
      extractNodeFields(e, r, o.get(t));
    }
  }
};

var gc = () => {
  if (w.optimisticOrder.length) {
    return;
  }
  for (var e of w.gc.keys()) {
    w.gc.delete(e);
    if ((w.refCount.get(e) || 0) > 0) {
      continue;
    }
    w.refCount.delete(e);
    w.records.base.delete(e);
    var r = w.links.base.get(e);
    if (r) {
      w.links.base.delete(e);
      for (var t in r) {
        updateRCForLink(r[t], -1);
      }
    }
  }
};

var updateDependencies = (e, r) => {
  if (e !== w.queryRootKey) {
    x.add(e);
  } else if (void 0 !== r && "__typename" !== r) {
    x.add(joinKeys(e, r));
  }
};

var updatePersist = (e, r) => {
  if (!I && w.storage) {
    w.persist.add(((e, r) => `${e.replace(/\./g, "%2e")}.${r}`)(e, r));
  }
};

var readRecord = (e, r) => {
  updateDependencies(e, r);
  return getNode(w.records, e, r);
};

var readLink = (e, r) => {
  updateDependencies(e, r);
  return getNode(w.links, e, r);
};

var writeRecord = (e, r, t) => {
  updateDependencies(e, r);
  updatePersist(e, r);
  setNode(w.records, e, r, t);
};

var hasField = (e, r) => void 0 !== readRecord(e, r) || void 0 !== readLink(e, r);

var writeLink = (e, r, t) => {
  var a = b ? w.links.optimistic.get(b) : w.links.base;
  if (!b) {
    var i = a && a.get(e);
    updateRCForLink(i && i[r], -1);
    updateRCForLink(t, 1);
  }
  updateDependencies(e, r);
  updatePersist(e, r);
  setNode(w.links, e, r, t);
};

var reserveLayer = (e, r, t) => {
  var a = e.optimisticOrder.indexOf(r);
  if (a > -1) {
    e.optimisticOrder.splice(a, 1);
  }
  if (t) {
    e.deferredKeys.add(r);
    for (a = a > -1 ? a : 0; a < e.optimisticOrder.length && !e.deferredKeys.has(e.optimisticOrder[a]) && (!e.dirtyKeys.has(e.optimisticOrder[a]) || !e.commutativeKeys.has(e.optimisticOrder[a])); a++) {}
  } else {
    e.deferredKeys.delete(r);
    if (a > -1 && !e.commutativeKeys.has(r)) {
      clearLayer(e, r);
    }
    a = 0;
  }
  e.optimisticOrder.splice(a, 0, r);
  e.commutativeKeys.add(r);
};

var createLayer = (e, r) => {
  if (-1 === e.optimisticOrder.indexOf(r)) {
    e.optimisticOrder.unshift(r);
  }
  if (!e.dirtyKeys.has(r)) {
    e.dirtyKeys.add(r);
    e.links.optimistic.set(r, new Map);
    e.records.optimistic.set(r, new Map);
  }
};

var clearLayer = (e, r) => {
  if (e.dirtyKeys.has(r)) {
    e.dirtyKeys.delete(r);
    e.records.optimistic.delete(r);
    e.links.optimistic.delete(r);
    e.deferredKeys.delete(r);
  }
};

var deleteLayer = (e, r) => {
  var t = e.optimisticOrder.indexOf(r);
  if (t > -1) {
    e.optimisticOrder.splice(t, 1);
    e.commutativeKeys.delete(r);
  }
  clearLayer(e, r);
};

var squashLayer = e => {
  var r = x;
  x = new Set;
  D = "write";
  var t = w.links.optimistic.get(e);
  if (t) {
    for (var a of t.entries()) {
      var i = a[0];
      var n = a[1];
      for (var o in n) {
        writeLink(i, o, n[o]);
      }
    }
  }
  var s = w.records.optimistic.get(e);
  if (s) {
    for (var v of s.entries()) {
      var l = v[0];
      var d = v[1];
      for (var u in d) {
        writeRecord(l, u, d[u]);
      }
    }
  }
  x = r;
  deleteLayer(w, e);
};

var inspectFields = e => {
  var {links: r, records: t} = w;
  var a = [];
  var i = new Set;
  updateDependencies(e);
  extractNodeMapFields(a, i, e, r);
  extractNodeMapFields(a, i, e, t);
  return a;
};

var persistData = () => {
  if (w.storage) {
    I = !0;
    D = "read";
    var r = {};
    for (var t of w.persist.keys()) {
      var {entityKey: a, fieldKey: i} = deserializeKeyInfo(t);
      var n = void 0;
      if (void 0 !== (n = readLink(a, i))) {
        r[t] = `:${e(n)}`;
      } else if (void 0 !== (n = readRecord(a, i))) {
        r[t] = e(n);
      } else {
        r[t] = void 0;
      }
    }
    I = !1;
    w.storage.writeData(r);
    w.persist.clear();
  }
};

var q = null;

var T = !1;

var getFieldError = e => e.__internal.path.length > 0 && e.__internal.errorMap ? e.__internal.errorMap[e.__internal.path.join(".")] : void 0;

var makeContext = (e, r, t, a, i, n) => {
  var o = {
    store: e,
    variables: r,
    fragments: t,
    parent: {
      __typename: a
    },
    parentTypeName: a,
    parentKey: i,
    parentFieldKey: "",
    fieldName: "",
    error: void 0,
    partial: !1,
    hasNext: !1,
    optimistic: I,
    __internal: {
      path: [],
      errorMap: void 0
    }
  };
  if (n && n.graphQLErrors) {
    for (var s = 0; s < n.graphQLErrors.length; s++) {
      var v = n.graphQLErrors[s];
      if (v.path && v.path.length) {
        if (!o.__internal.errorMap) {
          o.__internal.errorMap = Object.create(null);
        }
        o.__internal.errorMap[v.path.join(".")] = v;
      }
    }
  }
  return o;
};

var updateContext = (e, r, t, a, i, n) => {
  q = e;
  e.parent = r;
  e.parentTypeName = t;
  e.parentKey = a;
  e.parentFieldKey = i;
  e.fieldName = n;
  e.error = getFieldError(e);
};

var isFragmentHeuristicallyMatching = (e, r, t, a) => {
  if (!r) {
    return !1;
  }
  var i = getTypeCondition(e);
  if (!i || r === i) {
    return !0;
  }
  "production" !== process.env.NODE_ENV && warn("Heuristic Fragment Matching: A fragment is trying to match against the `" + r + "` type, but the type condition is `" + i + "`. Since GraphQL allows for interfaces `" + i + "` may be an interface.\nA schema needs to be defined for this match to be deterministic, otherwise the fragment will be matched heuristically!", 16);
  return "write" === D || !getSelectionSet(e).some((e => {
    if (e.kind !== n.FIELD) {
      return !1;
    }
    var r = keyOfField(getName(e), getFieldArguments(e, a));
    return !hasField(t, r);
  }));
};

var makeSelectionIterator = (e, r, t, a, i) => {
  var o;
  var s = 0;
  return function next() {
    var v;
    while (o || s < a.length) {
      v = void 0;
      T = t;
      if (o) {
        if (v = o()) {
          return v;
        } else {
          o = void 0;
          if ("production" !== process.env.NODE_ENV) {
            popDebugNode();
          }
        }
      } else {
        var l = a[s++];
        if (!shouldInclude(l, i.variables)) {} else if (l.kind !== n.FIELD) {
          var d = l.kind !== n.INLINE_FRAGMENT ? i.fragments[getName(l)] : l;
          if (d) {
            if (!d.typeCondition || (i.store.schema ? isInterfaceOfType(i.store.schema, d, e) : isFragmentHeuristicallyMatching(d, e, r, i.variables))) {
              if ("production" !== process.env.NODE_ENV) {
                pushDebugNode(e, d);
              }
              o = makeSelectionIterator(e, r, t || isDeferred(l, i.variables), getSelectionSet(d), i);
            }
          }
        } else if ("write" === D || !l._generated) {
          return l;
        }
      }
    }
  };
};

var ensureData = e => null == e ? null : e;

var ensureLink = (e, r) => {
  if (!r) {
    return r || null;
  } else if (Array.isArray(r)) {
    var t = new Array(r.length);
    for (var a = 0, i = t.length; a < i; a++) {
      t[a] = ensureLink(e, r[a]);
    }
    return t;
  }
  var n = e.keyOfEntity(r);
  if ("production" !== process.env.NODE_ENV) {
    if (!n && r && "object" == typeof r) {
      warn("Can't generate a key for link(...) item.\nYou have to pass an `id` or `_id` field or create a custom `keys` config for `" + r.__typename + "`.", 12);
    }
  }
  return n;
};

var _query = (e, t, a, i) => {
  var n = r(t.query);
  var o = getMainOperation(n);
  var s = e.rootFields[o.operation];
  var v = getSelectionSet(o);
  var l = makeContext(e, normalizeVariables(o, t.variables), getFragments(n), s, s, i);
  if ("production" !== process.env.NODE_ENV) {
    pushDebugNode(s, o);
  }
  var d = s !== l.store.rootFields.query ? readRoot(l, s, v, a || makeData()) : readSelection(l, s, v, a || makeData());
  if ("production" !== process.env.NODE_ENV) {
    popDebugNode();
    getCurrentDependencies();
  }
  return {
    dependencies: x,
    partial: l.partial || !d,
    hasNext: l.hasNext,
    data: d || null
  };
};

var readRoot = (e, r, t, a) => {
  if ("string" != typeof (e.store.rootNames[r] ? r : a.__typename)) {
    return a;
  }
  var i = makeSelectionIterator(r, r, T, t, e);
  var n;
  var o = V;
  var s = makeData(a);
  while (n = i()) {
    var v = getFieldAlias(n);
    var l = a[v];
    e.__internal.path.push(v);
    var d = void 0;
    if (n.selectionSet && null !== l) {
      d = readRootField(e, getSelectionSet(n), ensureData(l));
    } else {
      d = l;
    }
    o = o || d !== l;
    if (void 0 !== d) {
      s[v] = d;
    }
    e.__internal.path.pop();
  }
  return o ? s : a;
};

var readRootField = (e, r, t) => {
  if (Array.isArray(t)) {
    var a = new Array(t.length);
    var i = V;
    for (var n = 0, o = t.length; n < o; n++) {
      e.__internal.path.push(n);
      a[n] = readRootField(e, r, t[n]);
      i = i || a[n] !== t[n];
      e.__internal.path.pop();
    }
    return i ? a : t;
  } else if (null === t) {
    return null;
  }
  var s = e.store.keyOfEntity(t);
  if (null !== s) {
    return readSelection(e, s, r, t) || null;
  } else {
    return readRoot(e, t.__typename, r, t);
  }
};

function getFieldResolver(e, r, t, a) {
  var i = a.store.resolvers[r];
  var n = i && i[t];
  var o;
  for (var s in e) {
    var v = e[s];
    if (v && "include" !== s && "skip" !== s && a.store.directives[s]) {
      o = a.store.directives[s](getFieldArguments(v, a.variables));
      if ("production" === process.env.NODE_ENV) {
        return o;
      }
      break;
    }
  }
  if ("production" !== process.env.NODE_ENV) {
    if (n && o) {
      warn(`A resolver and directive is being used at "${r}.${t}" simultaneously. Only the directive will apply.`, 28);
    }
  }
  return o || n;
}

var readSelection = (e, r, t, a, i) => {
  var {store: n} = e;
  var o = r === n.rootFields.query;
  var s = i && n.keyOfEntity(i) || r;
  if ("production" !== process.env.NODE_ENV) {
    if (!o && e.store.rootNames[s]) {
      warn("Invalid root traversal: A selection was being read on `" + s + "` which is an uncached root type.\nThe `" + e.store.rootFields.mutation + "` and `" + e.store.rootFields.subscription + "` types are special Operation Root Types and cannot be read back from the cache.", 25);
    }
  }
  var v = !o ? readRecord(s, "__typename") || i && i.__typename : r;
  if ("string" != typeof v) {
    return;
  } else if (i && v !== i.__typename) {
    "production" !== process.env.NODE_ENV && warn("Invalid resolver data: The resolver at `" + s + "` returned an invalid typename that could not be reconciled with the cache.", 8);
    return;
  }
  var l = makeSelectionIterator(v, s, T, t, e);
  var d = !1;
  var u = !1;
  var p = V;
  var c;
  var f = e.partial;
  var y = makeData(a);
  while (void 0 !== (c = l())) {
    var h = getName(c);
    var m = getFieldArguments(c, e.variables);
    var g = getFieldAlias(c);
    var _ = getDirectives(c);
    var N = getFieldResolver(_, v, h, e);
    var k = keyOfField(h, m);
    var E = joinKeys(s, k);
    var O = readRecord(s, k);
    var w = i ? i[h] : void 0;
    if ("production" !== process.env.NODE_ENV && n.schema && v) {
      isFieldAvailableOnType(n.schema, v, h);
    }
    e.__internal.path.push(g);
    var b = void 0;
    if ("__typename" === h) {
      b = v;
    } else if (void 0 !== w && void 0 === c.selectionSet) {
      b = w;
    } else if ("read" === D && N) {
      var x = y;
      if (void 0 === c.selectionSet && void 0 !== O) {
        x = {
          ...y,
          [g]: O,
          [h]: O
        };
      }
      updateContext(e, x, v, s, k, h);
      b = N(x, m || {}, n, e);
      if (c.selectionSet) {
        b = resolveResolverResult(e, v, h, E, getSelectionSet(c), void 0 !== y[g] ? y[g] : a[g], b, ownsData(a));
      }
      if (n.schema && null === b && !isFieldNullable(n.schema, v, h)) {
        return;
      }
    } else if (!c.selectionSet) {
      b = O;
    } else if (void 0 !== w) {
      b = resolveResolverResult(e, v, h, E, getSelectionSet(c), void 0 !== y[g] ? y[g] : a[g], w, ownsData(a));
    } else {
      var I = readLink(s, k);
      if (void 0 !== I) {
        b = resolveLink(e, I, v, h, getSelectionSet(c), void 0 !== y[g] ? y[g] : a[g], ownsData(a));
      } else if ("object" == typeof O && null !== O) {
        b = O;
      }
    }
    if (!T && void 0 === b && (_.optional || getFieldError(e) || n.schema && isFieldNullable(n.schema, v, h))) {
      e.partial = !0;
      b = null;
    } else if (null === b && _.required) {
      b = void 0;
    } else {
      d = d || "__typename" !== h;
    }
    e.__internal.path.pop();
    p = p || b !== a[g];
    if (void 0 !== b) {
      y[g] = b;
    } else if (T) {
      u = !0;
    } else {
      e.partial = f;
      return;
    }
  }
  e.partial = e.partial || f;
  e.hasNext = e.hasNext || u;
  return o && e.partial && !d ? void 0 : p ? y : a;
};

var resolveResolverResult = (e, r, t, a, i, n, o, s) => {
  if (Array.isArray(o)) {
    var {store: v} = e;
    var l = v.schema ? isListNullable(v.schema, r, t) : !1;
    var d = e.partial;
    var u = makeData(n, !0);
    var p = V || !Array.isArray(n) || o.length !== n.length;
    for (var c = 0, f = o.length; c < f; c++) {
      e.__internal.path.push(c);
      var y = resolveResolverResult(e, r, t, joinKeys(a, `${c}`), i, null != n ? n[c] : void 0, o[c], s);
      e.__internal.path.pop();
      if (void 0 === y && !l) {
        e.partial = d;
        return;
      } else {
        e.partial = e.partial || void 0 === y && l;
        u[c] = null != y ? y : null;
        p = p || u[c] !== n[c];
      }
    }
    return p ? u : n;
  } else if (null == o) {
    return o;
  } else if (!s && null === n) {
    return null;
  } else if (isDataOrKey(o)) {
    var h = n || makeData(n);
    return "string" == typeof o ? readSelection(e, o, i, h) : readSelection(e, a, i, h, o);
  } else {
    "production" !== process.env.NODE_ENV && warn("Invalid resolver value: The field at `" + a + "` is a scalar (number, boolean, etc), but the GraphQL query expects a selection set for this field.", 9);
    return;
  }
};

var resolveLink = (e, r, t, a, i, n, o) => {
  if (Array.isArray(r)) {
    var {store: s} = e;
    var v = s.schema ? isListNullable(s.schema, t, a) : !1;
    var l = makeData(n, !0);
    var d = e.partial;
    var u = V || !Array.isArray(n) || r.length !== n.length;
    for (var p = 0, c = r.length; p < c; p++) {
      e.__internal.path.push(p);
      var f = resolveLink(e, r[p], t, a, i, null != n ? n[p] : void 0, o);
      e.__internal.path.pop();
      if (void 0 === f && !v) {
        e.partial = d;
        return;
      } else {
        e.partial = e.partial || void 0 === f && v;
        l[p] = f || null;
        u = u || l[p] !== n[p];
      }
    }
    return u ? l : n;
  } else if (null === r || null === n && o) {
    return null;
  }
  return readSelection(e, r, i, n || makeData(n));
};

var isDataOrKey = e => "string" == typeof e || "object" == typeof e && "string" == typeof e.__typename;

var _write = (e, t, a, i) => {
  if ("production" !== process.env.NODE_ENV) {
    getCurrentDependencies();
  }
  var n = r(t.query);
  var o = getMainOperation(n);
  var s = {
    data: a || makeData(),
    dependencies: x
  };
  var v = e.rootFields[o.operation];
  var l = makeContext(e, normalizeVariables(o, t.variables), getFragments(n), v, v, i);
  if ("production" !== process.env.NODE_ENV) {
    pushDebugNode(v, o);
  }
  writeSelection(l, v, getSelectionSet(o), s.data);
  if ("production" !== process.env.NODE_ENV) {
    popDebugNode();
  }
  return s;
};

var writeSelection = (e, r, t, a) => {
  var i = e.store.rootNames[r] || "query";
  var n = !!e.store.rootNames[r];
  var o = n ? r : a.__typename;
  if (!o && r && e.optimistic) {
    o = readRecord(r, "__typename");
  }
  if (!o) {
    "production" !== process.env.NODE_ENV && warn("Couldn't find __typename when writing.\nIf you're writing to the cache manually have to pass a `__typename` property on each entity in your data.", 14);
    return;
  } else if (!n && r) {
    writeRecord(r, "__typename", o);
  }
  var s = e.store.updates[o];
  var v = makeSelectionIterator(o, r || o, T, t, e);
  var l;
  while (l = v()) {
    var d = getName(l);
    var u = getFieldArguments(l, e.variables);
    var p = keyOfField(d, u);
    var c = getFieldAlias(l);
    var f = a[e.optimistic ? d : c];
    if ("__typename" === d || void 0 === f && (T || e.optimistic && "query" === i)) {
      continue;
    }
    if ("production" !== process.env.NODE_ENV) {
      if (e.store.schema && o && "__typename" !== d) {
        isFieldAvailableOnType(e.store.schema, o, d);
      }
    }
    e.__internal.path.push(c);
    var y = void 0;
    if (e.optimistic && "mutation" === i) {
      if (!(y = e.store.optimisticMutations[d])) {
        continue;
      }
    } else if (e.optimistic && "function" == typeof f) {
      y = f;
    }
    if (y) {
      updateContext(e, a, o, r || o, p, d);
      f = ensureData(y(u || {}, e.store, e));
    }
    if (void 0 === f) {
      if ("production" !== process.env.NODE_ENV) {
        if (!r || !hasField(r, p) || e.optimistic && !readRecord(r, "__typename")) {
          "production" !== process.env.NODE_ENV && warn("Invalid undefined: The field at `" + p + "` is `undefined`, but the GraphQL query expects a " + (void 0 === l.selectionSet ? "scalar (number, boolean, etc)" : "selection set") + " for this field.", 13);
        }
      }
      continue;
    }
    if (l.selectionSet) {
      if (r && "query" === i) {
        var h = joinKeys(r, p);
        var m = writeField(e, getSelectionSet(l), ensureData(f), h, e.optimistic ? readLink(r || o, p) : void 0);
        writeLink(r || o, p, m);
      } else {
        writeField(e, getSelectionSet(l), ensureData(f));
      }
    } else if (r && "query" === i) {
      writeRecord(r || o, p, null !== f || !getFieldError(e) ? f : void 0);
    }
    var g = s && s[d];
    if (g) {
      updateContext(e, a, o, r || o, p, d);
      a[d] = f;
      g(a, u || {}, e.store, e);
    }
    e.__internal.path.pop();
  }
};

var A = /^__|PageInfo|(Connection|Edge)$/;

var writeField = (e, r, t, a, i) => {
  if (Array.isArray(t)) {
    var n = new Array(t.length);
    for (var o = 0, s = t.length; o < s; o++) {
      e.__internal.path.push(o);
      var v = a ? joinKeys(a, `${o}`) : void 0;
      var l = writeField(e, r, t[o], v, null != i ? i[o] : void 0);
      n[o] = l;
      e.__internal.path.pop();
    }
    return n;
  } else if (null === t) {
    return getFieldError(e) ? void 0 : null;
  }
  var d = e.store.keyOfEntity(t) || ("string" == typeof i ? i : null);
  var u = t.__typename;
  if ("production" !== process.env.NODE_ENV) {
    if (a && !e.store.keys[t.__typename] && null === d && "string" == typeof u && !A.test(u)) {
      warn("Invalid key: The GraphQL query at the field at `" + a + "` has a selection set, but no key could be generated for the data at this field.\nYou have to request `id` or `_id` fields for all selection sets or create a custom `keys` config for `" + u + "`.\nEntities without keys will be embedded directly on the parent entity. If this is intentional, create a `keys` config for `" + u + "` that always returns null.", 15);
    }
  }
  var p = d || a;
  writeSelection(e, p, r, t);
  return p || null;
};

class Store {
  constructor(e) {
    if (!e) {
      e = {};
    }
    this.resolvers = e.resolvers || {};
    this.directives = e.directives || {};
    this.optimisticMutations = e.optimistic || {};
    this.keys = e.keys || {};
    this.globalIDs = Array.isArray(e.globalIDs) ? new Set(e.globalIDs) : !!e.globalIDs;
    var r = "Query";
    var t = "Mutation";
    var a = "Subscription";
    if (e.schema) {
      var i = (({__schema: e}) => {
        var r = new Map;
        var buildNameMap = e => {
          var r;
          return () => {
            if (!r) {
              r = {};
              for (var t = 0; t < e.length; t++) {
                r[e[t].name] = e[t];
              }
            }
            return r;
          };
        };
        var buildType = e => {
          switch (e.kind) {
           case "OBJECT":
           case "INTERFACE":
            return {
              name: e.name,
              kind: e.kind,
              interfaces: buildNameMap(e.interfaces || []),
              fields: buildNameMap(e.fields.map((e => ({
                name: e.name,
                type: e.type,
                args: buildNameMap(e.args)
              }))))
            };

           case "UNION":
            return {
              name: e.name,
              kind: e.kind,
              types: buildNameMap(e.possibleTypes || [])
            };
          }
        };
        var t = {
          query: e.queryType ? e.queryType.name : null,
          mutation: e.mutationType ? e.mutationType.name : null,
          subscription: e.subscriptionType ? e.subscriptionType.name : null,
          types: void 0,
          isSubType(e, t) {
            var a = r.get(e);
            var i = r.get(t);
            if (!a || !i) {
              return !1;
            } else if ("UNION" === a.kind) {
              return !!a.types()[t];
            } else if ("OBJECT" !== a.kind && "OBJECT" === i.kind) {
              return !!i.interfaces()[e];
            } else {
              return e === t;
            }
          }
        };
        if (e.types) {
          t.types = r;
          for (var a = 0; a < e.types.length; a++) {
            var i = e.types[a];
            if (i && i.name) {
              var n = buildType(i);
              if (n) {
                r.set(i.name, n);
              }
            }
          }
        }
        return t;
      })(e.schema);
      r = i.query || r;
      t = i.mutation || t;
      a = i.subscription || a;
      if (i.types) {
        this.schema = i;
      }
    }
    this.updates = e.updates || {};
    this.rootFields = {
      query: r,
      mutation: t,
      subscription: a
    };
    this.rootNames = {
      [r]: "query",
      [t]: "mutation",
      [a]: "subscription"
    };
    this.data = (n = r, {
      hydrating: !1,
      defer: !1,
      gc: new Set,
      persist: new Set,
      queryRootKey: n,
      refCount: new Map,
      links: {
        optimistic: new Map,
        base: new Map
      },
      records: {
        optimistic: new Map,
        base: new Map
      },
      deferredKeys: new Set,
      commutativeKeys: new Set,
      dirtyKeys: new Set,
      optimisticOrder: [],
      storage: null
    });
    var n;
    if (this.schema && "production" !== process.env.NODE_ENV) {
      !function expectValidKeyingConfig(e, r) {
        if ("production" !== process.env.NODE_ENV) {
          for (var t in r) {
            if ("production" !== process.env.NODE_ENV) {
              if (!e.types.has(t)) {
                warn("Invalid Object type: The type `" + t + "` is not an object in the defined schema, but the `keys` option is referencing it.", 20);
              }
            }
          }
        }
      }(this.schema, this.keys);
      !function expectValidUpdatesConfig(e, r) {
        if ("production" === process.env.NODE_ENV) {
          return;
        }
        for (var t in r) {
          if (!r[t]) {
            continue;
          } else if (!e.types.has(t)) {
            var a = "";
            if ("Mutation" === t && e.mutation && "Mutation" !== e.mutation) {
              a += "\nMaybe your config should reference `" + e.mutation + "`?";
            } else if ("Subscription" === t && e.subscription && "Subscription" !== e.subscription) {
              a += "\nMaybe your config should reference `" + e.subscription + "`?";
            }
            return "production" !== process.env.NODE_ENV ? warn("Invalid updates type: The type `" + t + "` is not an object in the defined schema, but the `updates` config is referencing it." + a, 21) : void 0;
          }
          var i = e.types.get(t).fields();
          for (var n in r[t]) {
            if ("production" !== process.env.NODE_ENV) {
              if (!i[n]) {
                warn("Invalid updates field: `" + n + "` on `" + t + "` is not in the defined schema, but the `updates` config is referencing it.", 22);
              }
            }
          }
        }
      }(this.schema, this.updates);
      !function expectValidResolversConfig(e, r) {
        if ("production" === process.env.NODE_ENV) {
          return;
        }
        for (var t in r) {
          if ("Query" === t) {
            if (e.query) {
              var a = e.types.get(e.query).fields();
              for (var i in r.Query || {}) {
                if (!a[i]) {
                  warnAboutResolver("Query." + i);
                }
              }
            } else {
              warnAboutResolver("Query");
            }
          } else if (!e.types.has(t)) {
            warnAboutResolver(t);
          } else if ("INTERFACE" === e.types.get(t).kind || "UNION" === e.types.get(t).kind) {
            s = t, v = e.types.get(t).kind, "production" !== process.env.NODE_ENV && warn(`Invalid resolver: \`${s}\` does not match to a concrete type in the schema, but the \`resolvers\` option is referencing it. Implement the resolver for the types that ${"UNION" === v ? "make up the union" : "implement the interface"} instead.`, 26);
          } else {
            var n = e.types.get(t).fields();
            for (var o in r[t] || {}) {
              if (!n[o]) {
                warnAboutResolver(t + "." + o);
              }
            }
          }
        }
        var s, v;
      }(this.schema, this.resolvers);
      !function expectValidOptimisticMutationsConfig(e, r) {
        if ("production" === process.env.NODE_ENV) {
          return;
        }
        if (e.mutation) {
          var t = e.types.get(e.mutation).fields();
          for (var a in r) {
            if ("production" !== process.env.NODE_ENV) {
              if (!t[a]) {
                warn(`Invalid optimistic mutation field: \`${a}\` is not a mutation field in the defined schema, but the \`optimistic\` option is referencing it.`, 24);
              }
            }
          }
        }
      }(this.schema, this.optimisticMutations);
    }
  }
  keyOfField(e, r) {
    return keyOfField(e, r);
  }
  keyOfEntity(e) {
    if (q && e === q.parent) {
      return q.parentKey;
    } else if (null == e || "string" == typeof e) {
      return e || null;
    } else if (!e.__typename) {
      return null;
    } else if (this.rootNames[e.__typename]) {
      return e.__typename;
    }
    var r = null;
    if (this.keys[e.__typename]) {
      r = this.keys[e.__typename](e) || null;
    } else if (null != e.id) {
      r = `${e.id}`;
    } else if (null != e._id) {
      r = `${e._id}`;
    }
    var t = e.__typename;
    return !0 === this.globalIDs || this.globalIDs && this.globalIDs.has(t) || !r ? r : `${t}:${r}`;
  }
  resolve(e, r, t) {
    var a = null;
    var i = this.keyOfEntity(e);
    if (i) {
      var n = keyOfField(r, t);
      if (void 0 === (a = readRecord(i, n))) {
        a = readLink(i, n);
      }
    }
    return a;
  }
  resolveFieldByKey(e, r, t) {
    return this.resolve(e, r, t);
  }
  invalidate(e, r, t) {
    var a = this.keyOfEntity(e);
    invariant(a, "production" !== process.env.NODE_ENV ? "Can't generate a key for invalidate(...).\nYou have to pass an id or _id field or create a custom `keys` field for `" + ("object" == typeof e ? e.__typename : e + "`.") : "", 19);
    ((e, r, t) => {
      var a = r ? [ {
        fieldKey: keyOfField(r, t)
      } ] : inspectFields(e);
      for (var i = 0, n = a.length; i < n; i++) {
        var {fieldKey: o} = a[i];
        if (void 0 !== readLink(e, o)) {
          writeLink(e, o, void 0);
        } else {
          writeRecord(e, o, void 0);
        }
      }
    })(a, r, t);
  }
  inspectFields(e) {
    var r = this.keyOfEntity(e);
    return r ? inspectFields(r) : [];
  }
  updateQuery(e, r) {
    var a = t(e.query, e.variables);
    var i = r(this.readQuery(a));
    if (null !== i) {
      _write(this, a, i, void 0);
    }
  }
  readQuery(e) {
    var r = t(e.query, e.variables);
    return _query(this, r, void 0, void 0).data;
  }
  readFragment(e, t, a, i) {
    return ((e, r, t, a, i) => {
      var n = getFragments(r);
      var o;
      if (i) {
        if (!(o = n[i])) {
          "production" !== process.env.NODE_ENV && warn("readFragment(...) was called with a fragment name that does not exist.\nYou provided " + i + " but could only find " + Object.keys(n).join(", ") + ".", 6);
          return null;
        }
      } else if (!(o = n[Object.keys(n)[0]])) {
        "production" !== process.env.NODE_ENV && warn("readFragment(...) was called with an empty fragment.\nYou have to call it with at least one fragment in your GraphQL document.", 6);
        return null;
      }
      var s = getFragmentTypeName(o);
      if ("string" != typeof t && !t.__typename) {
        t.__typename = s;
      }
      var v = e.keyOfEntity(t);
      if (!v) {
        "production" !== process.env.NODE_ENV && warn("Can't generate a key for readFragment(...).\nYou have to pass an `id` or `_id` field or create a custom `keys` config for `" + s + "`.", 7);
        return null;
      }
      if ("production" !== process.env.NODE_ENV) {
        pushDebugNode(s, o);
      }
      var l = makeContext(e, a || {}, n, s, v, void 0);
      var d = readSelection(l, v, getSelectionSet(o), makeData()) || null;
      if ("production" !== process.env.NODE_ENV) {
        popDebugNode();
      }
      return d;
    })(this, r(e), t, a, i);
  }
  writeFragment(e, t, a, i) {
    ((e, r, t, a, i) => {
      var n = getFragments(r);
      var o;
      if (i) {
        if (!(o = n[i])) {
          "production" !== process.env.NODE_ENV && warn("writeFragment(...) was called with a fragment name that does not exist.\nYou provided " + i + " but could only find " + Object.keys(n).join(", ") + ".", 11);
          return null;
        }
      } else if (!(o = n[Object.keys(n)[0]])) {
        "production" !== process.env.NODE_ENV && warn("writeFragment(...) was called with an empty fragment.\nYou have to call it with at least one fragment in your GraphQL document.", 11);
        return null;
      }
      var s = getFragmentTypeName(o);
      var v = {
        __typename: s,
        ...t
      };
      var l = e.keyOfEntity(v);
      if (!l) {
        return "production" !== process.env.NODE_ENV ? warn("Can't generate a key for writeFragment(...) data.\nYou have to pass an `id` or `_id` field or create a custom `keys` config for `" + s + "`.", 12) : void 0;
      }
      if ("production" !== process.env.NODE_ENV) {
        pushDebugNode(s, o);
      }
      var d = makeContext(e, a || {}, n, s, l, void 0);
      writeSelection(d, l, getSelectionSet(o), v);
      if ("production" !== process.env.NODE_ENV) {
        popDebugNode();
      }
    })(this, r(e), t, a, i);
  }
  link(e, r, ...t) {
    var a = 2 === t.length ? t[0] : null;
    var i = 2 === t.length ? t[1] : t[0];
    var n = this.keyOfEntity(e);
    if (n) {
      writeLink(n, keyOfField(r, a), ensureLink(this, i));
    }
  }
}

var addMetadata = (e, r) => a(e.kind, e, {
  ...e.context,
  meta: {
    ...e.context.meta,
    ...r
  }
});

var toRequestPolicy = (e, r) => a(e.kind, e, {
  ...e.context,
  requestPolicy: r
});

var cacheExchange = e => ({forward: t, client: i, dispatchDebug: n}) => {
  var o = new Store(e);
  if (e && e.storage) {
    o.data.hydrating = !0;
    e.storage.readData().then((r => {
      ((e, r, t) => {
        initDataState("write", e, null);
        for (var a in t) {
          var i = t[a];
          if (void 0 !== i) {
            var {entityKey: n, fieldKey: o} = deserializeKeyInfo(a);
            if (":" === i[0]) {
              if (void 0 === readLink(n, o)) {
                writeLink(n, o, JSON.parse(i.slice(1)));
              }
            } else if (void 0 === readRecord(n, o)) {
              writeRecord(n, o, JSON.parse(i));
            }
          }
        }
        e.storage = r;
        e.hydrating = !1;
        clearDataState();
      })(o.data, e.storage, r);
    }));
  }
  var f = new Map;
  var y = [];
  var h = new Map;
  var m = new Map;
  var g = new Set;
  var _ = new Set;
  var N = new Map;
  var k = new Set;
  var E = new Set;
  var isBlockedByOptimisticUpdate = e => {
    for (var r of e.values()) {
      if (g.has(r)) {
        return !0;
      }
    }
    return !1;
  };
  var collectPendingOperations = (e, r) => {
    if (r) {
      for (var t of r.values()) {
        var a = N.get(t);
        if (a) {
          for (var i of a.values()) {
            e.add(i);
          }
        }
      }
    }
  };
  var executePendingOperations = (e, r, t) => {
    for (var a of r.values()) {
      if (a !== e.key) {
        var n = h.get(a);
        if (n) {
          if ("query" === e.kind) {
            E.add(a);
          }
          var o = "cache-first";
          if (_.has(a)) {
            _.delete(a);
            o = "cache-and-network";
          }
          i.reexecuteOperation(toRequestPolicy(n, o));
        }
      }
    }
    if (!t) {
      var s = k;
      if ("query" === e.kind) {
        (k = E).add(e.key);
      }
      (E = s).clear();
    }
  };
  var prepareForwardedOperation = e => {
    var t = !1;
    if ("query" === e.kind) {
      reserveLayer(o.data, e.key);
      h.set(e.key, e);
    } else if ("teardown" === e.kind) {
      h.delete(e.key);
      m.delete(e.key);
      k.delete(e.key);
      noopDataState(o.data, e.key);
      return e;
    } else if ("mutation" === e.kind && "network-only" !== e.context.requestPolicy) {
      h.set(e.key, e);
      initDataState("write", o.data, e.key, !0, !1);
      var {dependencies: i} = _write(o, e, void 0, void 0);
      clearDataState();
      if (i.size) {
        for (var n of i.values()) {
          g.add(n);
        }
        f.set(e.key, i);
        var s = new Set;
        collectPendingOperations(s, i);
        executePendingOperations(e, s, !0);
        t = !0;
      }
    }
    return a(e.kind, {
      key: e.key,
      query: r(e.query),
      variables: e.variables ? filterVariables(getMainOperation(e.query), e.variables) : e.variables
    }, {
      ...e.context,
      optimistic: t
    });
  };
  var updateDependencies = (e, r) => {
    for (var t of r.values()) {
      var a = N.get(t);
      if (!a) {
        N.set(t, a = new Set);
      }
      a.add(e.key);
    }
  };
  var operationResultFromCache = e => {
    initDataState("read", o.data, void 0, !1, !1);
    var r = _query(o, e, m.get(e.key), void 0);
    clearDataState();
    var t = r.data ? !r.partial && !r.hasNext ? "hit" : "partial" : "miss";
    m.set(e.key, r.data);
    h.set(e.key, e);
    updateDependencies(e, r.dependencies);
    return {
      outcome: t,
      operation: e,
      data: r.data,
      dependencies: r.dependencies,
      hasNext: r.hasNext
    };
  };
  var updateCacheWithResult = (e, r) => {
    var t = h.get(e.operation.key) || e.operation;
    if ("mutation" === t.kind) {
      var a = f.get(t.key);
      collectPendingOperations(r, a);
      f.delete(t.key);
    }
    if ("subscription" === t.kind || e.hasNext) {
      reserveLayer(o.data, t.key, !0);
    }
    var i;
    var n = e.data;
    if (n) {
      initDataState("write", o.data, t.key, !1, !1);
      var s = _write(o, t, n, e.error).dependencies;
      clearDataState();
      collectPendingOperations(r, s);
      var v = "query" === t.kind ? m.get(t.key) : null;
      initDataState("read", o.data, t.key, !1, v !== n);
      var l = _query(o, t, v || n, e.error);
      clearDataState();
      n = l.data;
      if ("query" === t.kind) {
        collectPendingOperations(r, i = l.dependencies);
        m.set(t.key, n);
      }
    } else {
      noopDataState(o.data, t.key);
    }
    if (i) {
      updateDependencies(e.operation, i);
    }
    return {
      operation: t,
      data: n,
      error: e.error,
      extensions: e.extensions,
      hasNext: e.hasNext,
      stale: e.stale
    };
  };
  return e => {
    var r = s(v(operationResultFromCache)(l((e => "query" === e.kind && "network-only" !== e.context.requestPolicy))(e)));
    var a = l((e => "query" !== e.kind || "network-only" === e.context.requestPolicy))(e);
    var m = v((e => {
      "production" !== process.env.NODE_ENV && n({
        type: "cacheMiss",
        message: "The result could not be retrieved from the cache",
        operation: e.operation,
        source: "cacheExchange"
      });
      return "production" !== process.env.NODE_ENV ? addMetadata(e.operation, {
        cacheOutcome: "miss"
      }) : e.operation;
    }))(l((e => "miss" === e.outcome && "cache-only" !== e.operation.context.requestPolicy && !isBlockedByOptimisticUpdate(e.dependencies) && !k.has(e.operation.key)))(r));
    var N = v((e => {
      var {requestPolicy: r} = e.operation.context;
      var t = "cache-only" !== r && (e.hasNext || "cache-and-network" === r || "cache-first" === r && "partial" === e.outcome && !k.has(e.operation.key));
      var a = {
        operation: "production" !== process.env.NODE_ENV ? addMetadata(e.operation, {
          cacheOutcome: e.outcome
        }) : e.operation,
        data: e.data,
        error: e.error,
        extensions: e.extensions,
        stale: t && !e.hasNext,
        hasNext: t && e.hasNext
      };
      if (!t) {} else if (!isBlockedByOptimisticUpdate(e.dependencies)) {
        i.reexecuteOperation(toRequestPolicy(h.get(e.operation.key) || e.operation, "network-only"));
      } else if ("cache-and-network" === r) {
        _.add(e.operation.key);
      }
      "production" !== process.env.NODE_ENV && n({
        type: "cacheHit",
        message: "A requested operation was found and returned from the cache.",
        operation: e.operation,
        data: {
          value: a
        },
        source: "cacheExchange"
      });
      return a;
    }))(l((e => "miss" !== e.outcome || "cache-only" === e.operation.context.requestPolicy))(r));
    var E = t(v(prepareForwardedOperation)(d([ a, m ])));
    var O = v((e => {
      var r = new Set;
      var t = updateCacheWithResult(e, r);
      executePendingOperations(e.operation, r, !1);
      return t;
    }))(l((e => !f.has(e.operation.key)))(E));
    var w = u((e => {
      if (y.push(e) < f.size) {
        return p;
      }
      for (var r = 0; r < y.length; r++) {
        reserveLayer(o.data, y[r].operation.key);
      }
      g.clear();
      var t = [];
      var a = new Set;
      var i;
      while (i = y.shift()) {
        t.push(updateCacheWithResult(i, a));
      }
      executePendingOperations(e.operation, a, !1);
      return c(t);
    }))(l((e => f.has(e.operation.key)))(E));
    return d([ O, w, N ]);
  };
};

var K = {
  "cache-only": 0,
  "cache-first": 1,
  "network-only": 2,
  "cache-and-network": 3
};

var offlineExchange = e => r => {
  var {storage: n} = e;
  var o = e.isOfflineError || (e => e && e.networkError && !e.response && ("undefined" != typeof navigator && !1 === navigator.onLine || /request failed|failed to fetch|network\s?error/i.test(e.networkError.message)));
  if (n && n.onOnline && n.readMetadata && n.writeMetadata) {
    var {forward: v, client: u, dispatchDebug: p} = r;
    var {source: c, next: h} = f();
    var m = [];
    var g = !1;
    var _ = !1;
    var updateMetadata = () => {
      if (g) {
        var e = [];
        for (var r = 0; r < m.length; r++) {
          var t = m[r];
          if ("mutation" === t.kind) {
            e.push({
              query: i(t.query),
              variables: t.variables,
              extensions: t.extensions
            });
          }
        }
        n.writeMetadata(e);
      }
    };
    var filterQueue = e => {
      for (var r = m.length - 1; r >= 0; r--) {
        if (m[r].key === e) {
          m.splice(r, 1);
        }
      }
    };
    var flushQueue = () => {
      if (!_) {
        var e = new Set;
        _ = !0;
        for (var r = 0; r < m.length; r++) {
          var t = m[r];
          if ("mutation" === t.kind || !e.has(t.key)) {
            e.add(t.key);
            if ("subscription" !== t.kind) {
              h(a("teardown", t));
              var i = "cache-first";
              for (var n = 0; n < m.length; n++) {
                var {requestPolicy: o} = m[n].context;
                if (K[o] > K[i]) {
                  i = o;
                }
              }
              h(toRequestPolicy(t, i));
            } else {
              h(toRequestPolicy(t, "cache-first"));
            }
          }
        }
        _ = !1;
        m.length = 0;
        updateMetadata();
      }
    };
    var N = cacheExchange({
      ...e,
      storage: {
        ...n,
        readData() {
          var e = n.readData();
          return {
            async then(r) {
              var a = await n.readMetadata();
              for (var i = 0; a && i < a.length; i++) {
                m.push(u.createRequestOperation("mutation", t(a[i].query, a[i].variables), a[i].extensions));
              }
              r(await e);
              n.onOnline(flushQueue);
              g = !0;
              flushQueue();
            }
          };
        }
      }
    })({
      client: u,
      dispatchDebug: p,
      forward: e => s(l((e => {
        if (g && "mutation" === e.operation.kind && e.operation.context.optimistic && o(e.error, e)) {
          m.push(e.operation);
          updateMetadata();
          return !1;
        }
        return !0;
      }))(v(e)))
    });
    return e => {
      var r = d([ c, y((e => {
        if ("query" === e.kind && !g) {
          m.push(e);
        } else if ("teardown" === e.kind) {
          filterQueue(e.key);
        }
      }))(e) ]);
      return l((e => {
        if ("query" === e.operation.kind) {
          if (o(e.error, e)) {
            h(toRequestPolicy(e.operation, "cache-only"));
            m.push(e.operation);
            return !1;
          } else if (!g) {
            filterQueue(e.operation.key);
          }
        }
        return !0;
      }))(N(r));
    };
  }
  return cacheExchange(e)(r);
};

export { Store, cacheExchange, offlineExchange };
